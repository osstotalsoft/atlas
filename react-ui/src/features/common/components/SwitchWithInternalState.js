import React, { useCallback, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { FormControlLabel, Switch } from '@material-ui/core'
function SwitchWithInternalState({ checked, onChange, disabled, labelOff, labelOn }) {
  const [state, setState] = useState(checked)

  const handlePropertyChanged = useCallback(
    event => {
      setState(event.target.checked)
      onChange(event.target.checked)
    },
    [onChange]
  )

  useEffect(() => {
    setState(checked)
  }, [checked])

  return (
    <FormControlLabel
      control={<Switch checked={state} onChange={handlePropertyChanged} disabled={disabled} name='isOptional' color='secondary' />}
      label={state ? labelOn : labelOff}
    />
  )
}

SwitchWithInternalState.propTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  labelOff: PropTypes.string,
  labelOn: PropTypes.string
}

export default SwitchWithInternalState
