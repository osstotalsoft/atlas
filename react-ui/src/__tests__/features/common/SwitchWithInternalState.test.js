import React from 'react'
import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import SwitchWithInternalState from 'features/common/components/SwitchWithInternalState'
import renderer from 'react-test-renderer'

describe('Switch with internal state should work as expected', () => {
  it('Should render correctly', () => {
    const onClickMock = jest.fn()

    const tree = renderer.create(<SwitchWithInternalState checked={true} onChange={onClickMock} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render with no errors', () => {
    const onClickMock = jest.fn()

    const component = render(<SwitchWithInternalState checked={true} onChange={onClickMock} />)
    expect(component).toBeDefined()
  })

  it('Should display only the off label (TestOff)', () => {
    const onClickMock = jest.fn()

    const { queryByText } = render(
      <SwitchWithInternalState checked={false} onChange={onClickMock} labelOff={'TestOff'} labelOn={'TestOn'} />
    )
    expect(queryByText('TestOff')).toBeInTheDocument()
    expect(queryByText('TestOn')).not.toBeInTheDocument()
  })

  it('Should call onChange once for one click hit', () => {
    const onClickMock = jest.fn()
    const { getByRole } = render(
      <SwitchWithInternalState checked={false} onChange={onClickMock} labelOff={'TestOff'} labelOn={'TestOn'}></SwitchWithInternalState>
    )

    const toggle = getByRole('checkbox')
    fireEvent.click(toggle)
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  it('Should be secondary color when true .MuiSwitch-colorSecondary.Mui-checked', () => {
    const onClickMock = jest.fn()
    render(
      <SwitchWithInternalState checked={false} onChange={onClickMock} labelOff={'TestOff'} labelOn={'TestOn'}></SwitchWithInternalState>
    )
    const toggle = screen.getByRole('checkbox')
    fireEvent.click(toggle)
    const toggleOn = screen.getByRole('checkbox')
    //expect(toggleOn).toHaveClass('.MuiSwitch-colorSecondary.Mui-checked') TODO
  })
})
