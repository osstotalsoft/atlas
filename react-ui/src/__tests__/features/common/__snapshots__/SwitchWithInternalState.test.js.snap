// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch with internal state should work as expected Should render correctly 1`] = `
<label
  className="MuiFormControlLabel-root"
>
  <span
    className="MuiSwitch-root"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-2 Mui-checked"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          checked={true}
          className="PrivateSwitchBase-input-4 MuiSwitch-input"
          name="isOptional"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="MuiSwitch-thumb"
        />
      </span>
    </span>
    <span
      className="MuiSwitch-track"
    />
  </span>
  <span
    className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
  />
</label>
`;
